### PR TITLE
Adjust initialization in maximum and minimum (version 2)

### PR DESCRIPTION
**Pull request by andreasnoack**
See https://github.com/JuliaLang/julia/pull/27845 and https://github.com/JuliaLang/julia/issues/28101#issuecomment-404952538

### CODE REVIEW

**Comment by giordano** [comment]
Maybe add a test for #27992, too?

**Comment by andreasnoack** [comment]
> Maybe add a test for #27992, too?

I've just added that.

File: test/reducedim.jl [hunk 1]
```diff
@@ -50,14 +50,14 @@ safe_minabs(A::Array{T}, region) where {T} = safe_mapslices(minimum, abs.(A), re
...
-    @test sum(Areduc, dims=region) ≈ safe_sum(Areduc, region)
-    @test prod(Areduc, dims=region) ≈ safe_prod(Areduc, region)
-    @test maximum(Areduc, dims=region) ≈ safe_maximum(Areduc, region)
-    @test minimum(Areduc, dims=region) ≈ safe_minimum(Areduc, region)
-    @test sum(abs, Areduc, dims=region) ≈ safe_sumabs(Areduc, region)
-    @test sum(abs2, Areduc, dims=region) ≈ safe_sumabs2(Areduc, region)
-    @test maximum(abs, Areduc, dims=region) ≈ safe_maxabs(Areduc, region)
-    @test minimum(abs, Areduc, dims=region) ≈ safe_minabs(Areduc, region)
+    @test @inferred sum(Areduc, dims=region) ≈ safe_sum(Areduc, region)
+    @test @inferred prod(Areduc, dims=region) ≈ safe_prod(Areduc, region)
```
Comment (KristofferC) [inline test/reducedim.jl]: Does this `@inferred` apply to the approx operator or the LHS?

Comment (giordano) [reply test/reducedim.jl]: I guess it applies to the `isapprox`:
```julia
julia> f(x) = x < 0 ? '1' : 1
f (generic function with 1 method)

julia> using Test

julia> @inferred f(1)
ERROR: return type Int64 does not match inferred return type Union{Char, Int64}
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] top-level scope at none:0

julia> @inferred f(1) == 1
true

julia> @inferred(f(1)) == 1
ERROR: return type Int64 does not match inferred return type Union{Char, Int64}
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] top-level scope at none:0
```

File: base/abstractarray.jl [hunk 2]
```diff
@@ -1926,7 +1926,12 @@ function mapslices(f, A::AbstractArray; dims)
     Rsize = copy(dimsA)
     # TODO: maybe support removing dimensions
     if !isa(r1, AbstractArray) || ndims(r1) == 0
-        r1 = [r1]
+        # If the result of f on a single slice is a scalar then we add singleton
+        # dimensions. When adding adding the dimensions, we have to respect the
+        # index type of the input array (e.g. in the case of OffsetArrays)
+        tmp = similar(Aslice, typeof(r1), reduced_indices(Aslice, 1:ndims(Aslice)))
+        tmp[first(CartesianIndices(tmp))] = r1
```
Comment (nalimilan) [inline base/abstractarray.jl]: `firstindex(tmp)` would work too, right?

Typo above: "adding adding".

**Pull request merged**